src/memory/array.rs

```diff
@@ -0,0 +1,545 @@
+        self.set_width(width);
+        self
+    }
+
+    /// Returns the height of the ArrayDescripor
+    pub fn height(&self) -> usize {
+        self.desc.Height
+    }
+
+    /// Sets the height of the ArrayDescriptor
+    pub fn set_height(&mut self, height: usize) {
+        self.desc.Height = height;
+    }
+
+    /// Sets the height of the ArrayDescriptor. Composes with other with_ routines.
+    pub fn with_height(mut self, height: usize) -> Self {
+        self.set_height(height);
+        self
+    }
+
+    /// Returns the depth of the ArrayDescripor
+    pub fn depth(&self) -> usize {
+        self.desc.Depth
+    }
+
+    /// Sets the depth of the ArrayDescriptor
+    pub fn set_depth(&mut self, depth: usize) {
+        self.desc.Depth = depth;
+    }
+
+    /// Sets the depth of the ArrayDescriptor. Composes with other with_ routines.
+    pub fn with_depth(mut self, depth: usize) -> Self {
+        self.set_depth(depth);
+        self
+    }
+
+    /// Returns the format of the ArrayDescripor
+    pub fn format(&self) -> ArrayFormat {
+        ArrayFormat::from_raw(self.desc.Format)
+    }
+
+    /// Sets the format of the ArrayDescriptor
+    pub fn set_format(&mut self, format: ArrayFormat) {
+        self.desc.Format = format.to_raw();
+    }
+
+    /// Sets the format of the ArrayDescriptor. Composes with other with_ routines.
+    pub fn with_format(mut self, format: ArrayFormat) -> Self {
+        self.set_format(format);
+        self
+    }
+
+    /// Returns the number of channels in the ArrayDescriptor
+    pub fn num_channels(&self) -> c_uint {
+        self.desc.NumChannels
+    }
+
+    /// Sets the number of channels in the ArrayDescriptor
+    pub fn set_num_channels(&mut self, num_channels: c_uint) {
+        self.desc.NumChannels = num_channels;
+    }
+
+    /// Sets the number of channels in the ArrayDescriptor. Composes with other with_ routinese.
+    pub fn with_num_channels(mut self, num_channels: c_uint) -> Self {
+        self.set_num_channels(num_channels);
+        self
+    }
+
+    /// Returns the flags of the ArrayDescriptor
+    pub fn flags(&self) -> ArrayObjectFlags {
+        ArrayObjectFlags::from_bits_truncate(self.desc.Flags)
+    }
+
+    /// Sets the flags of the ArrayDescriptor.
+    pub fn set_flags(&mut self, flags: ArrayObjectFlags) {
+        self.desc.Flags = flags.bits();
+    }
+
+    /// Sets the flags of the ArrayDescriptor. Composes with other with_ routines.
+    pub fn with_flags(mut self, flags: ArrayObjectFlags) -> Self {
+        self.set_flags(flags);
+        self
+    }
+}
+
+/// A CUDA Array. Can be bound to a texture or surface.
+pub struct ArrayObject {
+    handle: CUarray,
+}
+
+impl ArrayObject {
+    /// Constructs a generic ArrayObject
+    pub fn from_descriptor(descriptor: ArrayDescriptor) -> CudaResult<Self> {
+        // We validate the descriptor up front. This provides a good error message to the user.
+        assert_ne!(
```

Comment:
Lets make these `debug_assert!`s and similar. As you noted, the CUDA driver will safely return an error if these checks are violated. It's great to add more-detailed error messages (and I think I might go back and add that to the other code) but it would also be nice to avoid runtime overhead in release builds.